tobler/util/util.py

```diff
@@ -124,6 +124,47 @@ def project_gdf(gdf, to_crs=None, to_latlong=False):
 def h3fy(source, resolution=6, clip=False, return_geoms=True):
     """Generate a hexgrid geodataframe that covers the face of a source geodataframe.
 
+    Parameters
+    ----------
+    source : geopandas.GeoDataFrame
+        GeoDataFrame to transform into a hexagonal grid
+    resolution : int, optional (default is 6)
+        resolution of output h3 hexgrid.
+        See <https://h3geo.org/docs/core-library/restable> for more information
+    clip : bool, optional (default is False)
+        if True, hexagons are clipped to the precise boundary of the source gdf. Otherwise,
+        heaxgons along the boundary will be left intact.
+    return_geoms: bool, optional (default is True)
+        whether to generate hexagon geometries as a geodataframe or simply return
+        hex ids as a pandas.Series
+
+    Returns
+    -------
+    pandas.Series or geopandas.GeoDataFrame
+        if `return_geoms` is True, a geopandas.GeoDataFrame whose rows comprise a hexagonal h3 grid (indexed on h3 hex id).
+        if `return_geoms` is False, a pandas.Series of h3 hexagon ids
+    """
+    # h3 hexes only work on polygons, not multipolygons
+    source = source.explode()
+
+    if type(source.unary_union) == Polygon:
```

Comment:
We do expensive `source.unary_union` here and then again in `_to_hex` or in the loop below. In larger dataframes, that can cause a significant slowdown. My sense is that you spend most of the time actually doing unary union than anything else in this function, so we surely don't want to do it twice.

Reply:
good point, thanks. I added a new PR that switches to a try/except which should cut down one `unary_union` call when it's not necessary, unless there's yet a more efficient way you can think of?

Reply:
For an approximate but faster solution, we could offer an option to simply cover the bounding box (or its alpha shape). That'd avoid the unary union and, in some cases, it should be the same.